File: app/src/main/java/com/manoj/dlt/ui/adapters/FilterableListAdapter.java
package com.manoj.dlt.ui.adapters;

import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Filter;
import android.widget.Filterable;

import java.util.ArrayList;
import java.util.List;

public abstract class FilterableListAdapter<T> extends BaseAdapter implements Filterable
{
    protected List<T> _resultList;
    protected List<T> _originalList;
    protected String _searchString;

    public FilterableListAdapter(List<T> originalList, boolean defaultListEmpty)
    {
        _resultList = new ArrayList<T>();
        if(!defaultListEmpty)
        {
            _resultList.addAll(originalList);
        }
        _originalList = originalList;
        _searchString = "";
    }

    @Override
    public int getCount()
    {
        Log.d("deep","size = "+_resultList.size());
        return _resultList.size();
    }

    @Override
    public Object getItem(int i)
    {
        Log.d("deep","call for item "+i);
        return _resultList.get(i);
    }

    public void updateBaseData(List<T> baseData)
    {
        _originalList = baseData;
        updateResults(_searchString);
        notifyDataSetChanged();
    }

    public void updateResults(CharSequence searchString)
    {
        getFilter().filter(searchString);
    }

    @Override
    public Filter getFilter()
    {
        return new Filter()
        {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence)
            {
                FilterResults results = new FilterResults();
                List<T> resultList = getMatchingResults(charSequence);
                results.values = resultList;
                results.count = resultList.size();
                Log.d("deep","filtering, cnt  = "+resultList.size());
                return results;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults)
            {
                _searchString = charSequence.toString();
                _resultList = (List<T>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }

    protected abstract List<T> getMatchingResults(CharSequence constraint);
}

File: app/src/main/kotlin/com/manoj/dlt/ui/adapters/FilterableListAdapter.kt
package com.manoj.dlt.ui.adapters

import android.util.Log
import android.widget.BaseAdapter
import android.widget.Filter
import android.widget.Filterable
import java.util.*

abstract class FilterableListAdapter<T>(protected var originalList: MutableList<T>, defaultListEmpty: Boolean) : BaseAdapter(), Filterable {
    protected var resultList = ArrayList<T>()
    protected var searchString = ""

    init {
        if (!defaultListEmpty) {
            resultList.addAll(originalList)
        }
    }

    override fun getCount(): Int {
        Log.d("deep", "size = " + resultList.size)
        return resultList.size
    }

    override fun getItem(i: Int): T {
        Log.d("deep", "call for item " + i)
        return resultList[i]
    }

    fun updateBaseData(baseData: List<T>) {
        originalList = baseData.toMutableList()
        updateResults(searchString)
    }

    fun updateResults(searchString: CharSequence) {
        filter.filter(searchString)
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): Filter.FilterResults {
                val results = Filter.FilterResults()
                val resultList = getMatchingResults(charSequence)
                results.values = resultList
                results.count = resultList.size
                Log.d("deep", "filtering, count  = " + resultList.size)
                return results
            }

            override fun publishResults(charSequence: CharSequence, filterResults: Filter.FilterResults) {
                searchString = charSequence.toString()
                resultList.clear()
                @Suppress("UNCHECKED_CAST")
                resultList.addAll(filterResults.values as List<T>)
                notifyDataSetChanged()
            }
        }
    }

    protected abstract fun getMatchingResults(constraint: CharSequence): List<T>
}
